syntax = "proto3";

package raft;

import "google/protobuf/empty.proto";

service Raft {
  rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
  rpc NotifyLeader(LeaderNotification) returns (google.protobuf.Empty);
  rpc Put(PutRequest) returns (PutResponse);
  rpc Replicate(ReplicateRequest) returns (ReplicateResponse);
  rpc WhoIsLeader(google.protobuf.Empty) returns (LeaderResponse);

}

message LogEntry {
  int32 index = 1;
  int32 term = 2;
  string key = 3;
  string value = 4;
}



message RequestVoteRequest {
  int32 term = 1;
  string candidate_id = 2;
}

message RequestVoteResponse {
  int32 term = 1;
  bool vote_granted = 2;
}

message AppendEntriesRequest {
  int32 term = 1;
  string leader_id = 2;
  int32 prev_log_index = 3;
  int32 prev_log_term = 4;
  repeated LogEntry entries = 5;
  int32 leader_commit = 6;
}

message AppendEntriesResponse {
  int32 term = 1;
  bool success = 2;
}

message LeaderNotification {
  string leader_id = 1;
}

message PutRequest {
  string key = 1;
  string value = 2;
}

message PutResponse {
  bool success = 1;
}

message ReplicateRequest {
  string key = 1;
  string value = 2;
}

message ReplicateResponse {
  bool ack = 1;
}

message LeaderResponse {
  string leader_id = 1;
  string address = 2;
}
